crates/ra_ide/src/completion/complete_attribute.rs

```diff
@@ -20,6 +20,14 @@ pub(super) fn complete_attribute(acc: &mut Completions, ctx: &CompletionContext)
         {
             complete_derive(acc, ctx, token_tree)
         }
+        (_, Some(ast::AttrInput::TokenTree(token_tree))) => {
+            let token_tree_str = token_tree.to_string();
```

Comment:
This needlessly allocates a string, and, as per our code style, this should be avoided: 

https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/README.md#premature-pessimization

Generally, unless you compute some genuinely new data, there's almost always a way to avoid allocation by carefully writing the code to process data in the form it currently is. 

In this case, the syntax tree obviously has information about parenthesis, so there has to be a way to extract that without allocating. The most straighforward one would be to add `.starts_with` and `.end_with` to SyntaxText, but it's also possible to look at the syntax structure itself (check that left_delimiter_token and right_delimiter_token are `T!['(']`).

However, just replacing this match arm with `()` make the tests pass, so we either don't need this at all, or we need a test, which shows why we need this logic.

Reply:
Indeed, fixed